### PR TITLE
Tidy Android Bridge

### PR DESCRIPTION
**Pull request by igstewart3**
This PR:
* Creates a new JsonConverter class and pulls the JSON conversion code out of the main module class
* Fixes build configuration issues that prevented JSON classes and bridging classes (ReadableMap etc.) being used in tests
* Removes thrown JSONExceptions from React Native methods as these cannot be caught on the JS level

### CODE REVIEW

File: android/src/main/java/com/sailthru/mobile/rnsdk/JsonConverter.kt [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+ * Class for handling conversion between Native Maps and Arrays and JSON Objects and Arrays.
+ */
+internal class JsonConverter {
+
+    /**
+     * Convert JSONObject to WritableMap
+     */
+    @Throws(JSONException::class)
+    fun convertJsonToMap(jsonObject: JSONObject): WritableMap {
+        val writableMap: WritableMap = createNativeMap()
```
Comment (dan-sailthru) [inline android/src/main/java/com/sailthru/mobile/rnsdk/JsonConverter.kt:19]: I'm curious if specifying the Type here is a stylistic choice? I don't think that kotlin requires this.

File: android/src/main/java/com/sailthru/mobile/rnsdk/JsonConverter.kt [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        val iterator = jsonObject.keys()
```
Comment (dan-sailthru) [inline android/src/main/java/com/sailthru/mobile/rnsdk/JsonConverter.kt:20]: you could save some boilerplate code here with something like

```
    fun convertJsonToMap1(jsonObject: JSONObject) = createNativeMap().apply {
        for(key in jsonObject.keys()) {
            when (val value = jsonObject[key]) {
                is JSONObject -> putMap(key, convertJsonToMap(value))
                ...
                else -> putString(key, value.toString())
            }
        }
    }
```

similarly for the other iterator-based methods in this class

File: android/src/main/java/com/sailthru/mobile/rnsdk/RNSailthruMobileModule.java [hunk 3]
```diff
@@ -510,21 +288,35 @@ public void onFailure(@NotNull Error error) {
...
-        if (typeCode == 0) type = ImpressionType.IMPRESSION_TYPE_IN_APP_VIEW;
-        else if (typeCode == 1) type = ImpressionType.IMPRESSION_TYPE_STREAM_VIEW;
-        else if (typeCode == 2) type = ImpressionType.IMPRESSION_TYPE_DETAIL_VIEW;
+        ImpressionType type;
+        switch (typeCode) {
+            case 0: type = ImpressionType.IMPRESSION_TYPE_IN_APP_VIEW; break;
+            case 1: type = ImpressionType.IMPRESSION_TYPE_STREAM_VIEW; break;
+            case 2: type = ImpressionType.IMPRESSION_TYPE_DETAIL_VIEW; break;
+            default: return;
+        }
```
Comment (dan-sailthru) [inline android/src/main/java/com/sailthru/mobile/rnsdk/RNSailthruMobileModule.java:297]: 🆒

File: android/src/test/java/com/sailthru/mobile/rnsdk/JsonConverterTest.kt [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+            put("test string")
+        }
+        val jsonObject = JSONObject().apply {
+            put(TEST_KEY, subArray)
+        }
+
+        val writableMap = jsonConverter.convertJsonToMap(jsonObject)
+
+        assertEquals("test string", writableMap.getArray(TEST_KEY)?.getString(0))
+    }
```
Comment (dan-sailthru) [inline android/src/test/java/com/sailthru/mobile/rnsdk/JsonConverterTest.kt:97]: is it possible to have a map of mixed types? I know JS allows this, but do these bridge classes? might be worth a test.

**Review by dan-sailthru — COMMENTED** [review]
Just a few kotlin things and I think we might be missing some test cases, but otherwise this is a solid refactor.

File: android/src/main/java/com/sailthru/mobile/rnsdk/JsonConverter.kt (hunk 1 above)
Comment (igstewart3) [reply android/src/main/java/com/sailthru/mobile/rnsdk/JsonConverter.kt:19]: I think that's left over from the Java -> Kotlin conversion, but you're right it is unnecessary. Will remove 👍

File: android/src/test/java/com/sailthru/mobile/rnsdk/JsonConverterTest.kt (hunk 4 above)
Comment (igstewart3) [reply android/src/test/java/com/sailthru/mobile/rnsdk/JsonConverterTest.kt:97]: Yeah, the map can contain whatever mix you like. Good shout, I'll add a more complicated test case 👍

**Review by dan-sailthru — APPROVED** [review]
looks good!

**Pull request merged**